components/ws-manager-mk2/controllers/workspace_controller.go

```diff
@@ -0,0 +1,876 @@
+	}
+
+	var (
+		initializer     csapi.WorkspaceInitializer
+		snc             wsdaemon.WorkspaceContentServiceClient
+		contentManifest []byte
+	)
+	// The function below deliniates the initializer lock. It's just there so that we can
+	// defer the unlock call, thus making sure we actually call it.
+	err = func() error {
+		r.initializerMapLock.Lock()
+		defer r.initializerMapLock.Unlock()
+
+		_, alreadyInitializing := r.initializerMap[ws.Name]
+		if alreadyInitializing {
+			return nil
+		}
+
+		// There is no need to emit this span if the operation is noop.
+		span, ctx := tracing.FromContext(ctx, "initializeWorkspace")
+		defer tracing.FinishSpan(span, &err)
+
+		err = proto.Unmarshal(ws.Spec.Initializer, &initializer)
+		if err != nil {
+			return xerrors.Errorf("cannot unmarshal init config: %w", err)
+		}
+
+		// connect to the appropriate ws-daemon
+		snc, err = r.connectToWorkspaceDaemon(ctx, ws)
+		if err != nil {
+			return err
+		}
+
+		// mark that we're already initialising this workspace
+		r.initializerMap[ws.Name] = struct{}{}
+
+		return nil
+	}()
+	if err != nil {
+		return xerrors.Errorf("cannot initialize workspace: %w", err)
+	}
+	if err == nil && snc == nil {
+		// we are already initialising
+		return nil
+	}
+	log.FromContext(ctx).Info("initialising workspace", "workspace", ws.Name)
+	_, err = snc.InitWorkspace(ctx, &wsdaemon.InitWorkspaceRequest{
+		Id: ws.Name,
+		Metadata: &wsdaemon.WorkspaceMetadata{
+			Owner:  ws.Spec.Ownership.Owner,
+			MetaId: ws.Spec.Ownership.WorkspaceID,
+		},
+		Initializer:           &initializer,
+		FullWorkspaceBackup:   false,
+		PersistentVolumeClaim: true,
+		ContentManifest:       contentManifest,
+		RemoteStorageDisabled: ws.Spec.Type == workspacev1.WorkspaceTypeImageBuild,
+	})
+	if st, ok := status.FromError(err); ok && st.Code() == codes.AlreadyExists {
+		// we're already initializing, things are good - we'll wait for it later
+		err = nil
+	}
+	if err != nil {
+		return xerrors.Errorf("cannot initialize workspace: %w", err)
+	}
+
+	return nil
+}
+
+var (
+	wsOwnerKey = ".metadata.controller"
+	apiGVStr   = workspacev1.GroupVersion.String()
+)
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *WorkspaceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	idx := func(rawObj client.Object) []string {
+		// grab the obj object, extract the owner...
+		obj := rawObj.(metav1.Object)
+		owner := metav1.GetControllerOf(obj)
+		if owner == nil {
+			return nil
+		}
+		// ...make sure it's a workspace...
+		if owner.APIVersion != apiGVStr || owner.Kind != "Workspace" {
+			return nil
+		}
+
+		// ...and if so, return it
+		return []string{owner.Name}
+	}
+	err := mgr.GetFieldIndexer().IndexField(context.Background(), &corev1.Pod{}, wsOwnerKey, idx)
+	if err != nil {
+		return err
+	}
+	err = mgr.GetFieldIndexer().IndexField(context.Background(), &corev1.PersistentVolumeClaim{}, wsOwnerKey, idx)
+	if err != nil {
+		return err
+	}
```

Comment:
I think we could remove this part, don't we?

Reply:
What is broken/unnecessary with this code?